modules/schema/src/main/java/org/apache/ignite/internal/schema/Row.java

```diff
@@ -17,167 +17,217 @@
-        public static final int TOMBSTONE = 1;
+    /** Binary row. */
+    private final BinaryRow row;
 
-        /** Null-value flag. */
-        public static final int NULL_VALUE = 1 << 1;
+    /**
+     * Constructor.
+     *
+     * @param schema Schema.
+     * @param row Binary row representation.
+     */
+    public Row(SchemaDescriptor schema, BinaryRow row) {
+        assert row.schemaVersion() == schema.version();
 
-        /** Stub. */
-        private RowFlags() {
-        }
+        this.row = row;
+        this.schema = schema;
     }
 
-    /** Schema descriptor for which this row was created. */
-    private final SchemaDescriptor schema;
-
     /**
-     * @param schema Schema instance.
+     * @return Row schema.
      */
-    protected Row(SchemaDescriptor schema) {
-        this.schema = schema;
+    public SchemaDescriptor rowSchema() {
+        return schema;
     }
 
     /**
      * @return {@code True} if row has non-null value, {@code false} otherwise.
      */
-    public boolean hasValue() {
-        short flags = readShort(FLAGS_FIELD_OFFSET);
-
-        return (flags & (RowFlags.NULL_VALUE | RowFlags.TOMBSTONE)) == 0;
+    @Override public boolean hasValue() {
+        return row.hasValue();
     }
 
     /**
+     * Reads value for specified column.
+     *
+     * @param col Column index.
+     * @return Column value.
+     * @throws InvalidTypeException If actual column type does not match the requested column type.
      */
-    public byte byteValue(int col) {
+    public byte byteValue(int col) throws InvalidTypeException {
         long off = findColumn(col, NativeTypeSpec.BYTE);
 
         return off < 0 ? 0 : readByte(offset(off));
     }
 
     /**
+     * Reads value for specified column.
+     *
+     * @param col Column index.
+     * @return Column value.
+     * @throws InvalidTypeException If actual column type does not match the requested column type.
      */
-    public Byte byteValueBoxed(int col) {
+    public Byte byteValueBoxed(int col) throws InvalidTypeException {
         long off = findColumn(col, NativeTypeSpec.BYTE);
 
         return off < 0 ? null : readByte(offset(off));
     }
 
     /**
+     * Reads value for specified column.
+     *
+     * @param col Column index.
+     * @return Column value.
+     * @throws InvalidTypeException If actual column type does not match the requested column type.
      */
-    public short shortValue(int col) {
+    public short shortValue(int col) throws InvalidTypeException {
         long off = findColumn(col, NativeTypeSpec.SHORT);
 
         return off < 0 ? 0 : readShort(offset(off));
     }
 
     /**
+     * Reads value for specified column.
+     *
+     * @param col Column index.
+     * @return Column value.
+     * @throws InvalidTypeException If actual column type does not match the requested column type.
      */
-    public Short shortValueBoxed(int col) {
+    public Short shortValueBoxed(int col) throws InvalidTypeException {
```

Comment:
"shortValue()" can't return 'null' but column can be NULLABLE.
But we can't leave "shortValueBoxed()" only to avoid unwanted boxing/unboxing in some scenarios.
E.g. to read a value from Row and write it directly to user POJO via reflection/VahHandle.